pkg/k8s/watchers/watcher.go

```diff
@@ -410,29 +409,11 @@ func (k *K8sWatcher) enableK8sWatchers(ctx context.Context, resourceNames []stri
 // K8sEventProcessed is called to do metrics accounting for each processed
 // Kubernetes event
 func (k *K8sWatcher) K8sEventProcessed(scope, action string, status bool) {
-	result := "success"
-	if !status {
-		result = "failed"
-	}
-
-	metrics.KubernetesEventProcessed.WithLabelValues(scope, action, result).Inc()
+	k.k8sEventReporter.K8sEventProcessed(scope, action, status)
```

Comment:
These methods an be completely avoided now right? If we just embed the event reporter:

```diff
diff --git a/pkg/k8s/watchers/watcher.go b/pkg/k8s/watchers/watcher.go
index e41f954b7d..6dc4487c70 100644
--- a/pkg/k8s/watchers/watcher.go
+++ b/pkg/k8s/watchers/watcher.go
@@ -129,7 +129,7 @@ type K8sWatcher struct {
 
 	clientset client.Clientset
 
-	k8sEventReporter *K8sEventReporter
+	*K8sEventReporter
 
 	// k8sResourceSynced maps a resource name to a channel. Once the given
 	// resource name is synchronized with k8s, the channel for which that
@@ -202,7 +202,7 @@ func newWatcher(
 		resourceGroupsFn:      resourceGroups,
 		db:                    db,
 		clientset:             clientset,
-		k8sEventReporter:      k8sEventReporter,
+		K8sEventReporter:      k8sEventReporter,
 		k8sResourceSynced:     k8sResourceSynced,
 		k8sAPIGroups:          k8sAPIGroups,
 		K8sSvcCache:           serviceCache,
@@ -405,15 +405,3 @@ func (k *K8sWatcher) enableK8sWatchers(ctx context.Context, resourceNames []stri
 
 	asyncControllers.Wait()
 }
-
-// K8sEventProcessed is called to do metrics accounting for each processed
-// Kubernetes event
-func (k *K8sWatcher) K8sEventProcessed(scope, action string, status bool) {
-	k.k8sEventReporter.K8sEventProcessed(scope, action, status)
-}
-
-// K8sEventReceived does metric accounting for each received Kubernetes event, as well
-// as notifying of events for k8s resources synced.
-func (k *K8sWatcher) K8sEventReceived(apiResourceName, scope, action string, valid, equal bool) {
-	k.k8sEventReporter.K8sEventReceived(apiResourceName, scope, action, valid, equal)
-}
```

Reply:
Yep that would be definitely possible and nicer. There are also a bunch of others that simply delegate to the function with the same name on fields: `GetCachedPod`, `GetCachedNamespace`, `GetAPIGroups`, `WaitForCacheSync`, ...

I kept these "delegation" methodson purpose to make the direct dependencies that are still pointing to the global `K8sWatcher` visible. IMO, inlining them would hide this fact. I plan to replace them with direct dependencies to the "inner" resources - in your mentioned case to the `K8sEventReporter` directly.

Reply:
Sounds good.